### PR TITLE
AdamW vs Lion: direct optimizer comparison at SOTA config (Round 19)

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Lion optimizer was selected early in the research programme (PR #411) as a replacement for AdamW. While Lion has proven superior in most subsequent experiments, a **direct head-to-head comparison at the current fully-tuned configuration has never been run**. All previous AdamW runs used sub-optimal architectures or hyperparameters.

This experiment establishes a definitive answer: given the full SOTA stack (STRING-sep pos encoding, QK-norm, vol-curriculum, tau loss weights, L=5 depth), does AdamW at an appropriately translated learning rate match or approach Lion's performance?

The standard LR translation from Lion to AdamW is ~10×, consistent with the historical transition in our BASELINE.md (AdamW lr≈1e-4 → Lion lr=9e-5 at the first switchover).

If AdamW at lr=9e-4 achieves EP4 val_abupt ≤ 6.5985%, it opens a new optimization axis (AdamW-specific regularization, decoupled weight decay tuning, β₂ scheduling). If it fails the EP4 gate, we confirm Lion is definitively optimal and close this hypothesis permanently.

## Instructions

Run a **4-epoch screen** using AdamW optimizer with the full SOTA config, swapping only the optimizer and LR:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse \
  --optimizer adamw \
  --lr 9e-4 \
  --weight-decay 0.01 \
  --tau-y-loss-weight 1.5 \
  --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.0 \
  --ema-decay 0.999 \
  --grad-clip-norm 0.5 \
  --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable \
  --use-qk-norm \
  --rff-num-features 16 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 \
  --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --wandb-group adamw-benchmark-r19 \
  --wandb-name alphonse/adamw-lr9e-4-wd1e-2
```

**Kill gates** (abort early if not met):
- EP1: val_abupt < 30%
- EP2: val_abupt < 16%
- EP3: val_abupt < 8%
- EP4: val_abupt ≤ 6.5985% → advance to 13-ep full run; if > 6.5985% → post results and close as CONFIRMED LION OPTIMAL

**If EP4 passes the gate**, run the full 13-epoch training:

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse \
  --optimizer adamw \
  --lr 9e-4 \
  --weight-decay 0.01 \
  --tau-y-loss-weight 1.5 \
  --tau-z-loss-weight 2.0 \
  --surface-loss-weight 2.0 \
  --ema-decay 0.999 \
  --grad-clip-norm 0.5 \
  --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable \
  --use-qk-norm \
  --rff-num-features 16 \
  --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 \
  --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 \
  --model-hidden-dim 512 \
  --model-heads 4 \
  --model-slices 128 \
  --batch-size 4 \
  --validation-every 1 \
  --train-surface-points 65536 \
  --eval-surface-points 65536 \
  --train-volume-points 65536 \
  --eval-volume-points 65536 \
  --wandb-group adamw-benchmark-r19 \
  --wandb-name alphonse/adamw-lr9e-4-wd1e-2-full
```

**Note:** `--no-compile-model` is mandatory — torch.compile + DDP causes NCCL deadlock at step 1.

**Report per-epoch val_abupt, val_surf_p, val_vol_p** in a PR comment as results come in.

## Baseline

Current single-model SOTA (PR #592, run `4k25s25e`):
- **val_abupt**: 6.5985% (target to beat)
- val_surf_p: 4.3322%
- val_vol_p: 3.9456%
- val_wall_x: 6.5420%
- val_wall_y: 8.3631%
- val_wall_z: 9.8099%
- test_abupt: 7.9915%

Ensemble SOTA (PR #612, K=7 greedy pool-24): val_abupt=6.1751%, test_abupt=7.5347%

Reproduce SOTA:
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Your PR has been open for over 3 hours with zero activity. Your pod is healthy (kubectl shows 1/1 Ready). Alphonse, you need to launch immediately.

**Your assignment:** PR #875 — AdamW vs Lion optimizer comparison. Run identical baseline config with AdamW replacing Lion to determine if Lion's advantage is genuine or artifact.

**Quick implementation:**
```python
# Replace Lion optimizer with AdamW:
optimizer = torch.optim.AdamW(
    model.parameters(), 
    lr=1e-3,           # AdamW typically needs higher LR than Lion
    weight_decay=5e-4,
    betas=(0.9, 0.999)
)
```
Keep everything else identical to the SOTA baseline (same vol_pts schedule 0:16384:1:32768:2:49152:3:65536, 4 epochs, same architecture).

**Run command:**
```bash
torchrun --standalone --nproc_per_node=8 train.py   --epochs 4   --optimizer adamw   --lr 1e-3   --wandb_group adamw-vs-lion-r18   --run_name alphonse/adamw-lr1e3-4ep
```

**Current baseline to beat:** val_abupt = 6.5985% (run `4k25s25e`, Lion optimizer)

**Kill gates:**
- EP1: <30%
- EP2: <16%
- EP3: <8%
- EP4: ≤6.5985%

**You have 15 minutes to post a launch confirmation with your W&B run ID, or tell me the specific blocker you are hitting.** Your GPU is sitting idle while other students are posting results.

**Comment by morganmcg1** [comment]
STUDENT alphonse: Launched. Apologies for the silent gap — the 4-epoch screen has been running since 20:18 UTC (now 22:18 UTC).

**Run details:**
- W&B rank-0 run ID: `thotfx2z`
- W&B URL: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/thotfx2z
- Group: `adamw-benchmark-r19`
- Optimizer: AdamW (default betas 0.9, 0.999), lr=9e-4, wd=0.01
- All other config matches PR spec (SOTA stack: STRING-sep, QK-norm, vol-curriculum, tau weights, L=5, 4ep)

**EP1 results (gate <30%): PASS**
- val_abupt: **22.345%**
- val_surf_p: 16.612%
- val_vol_p: 13.282%
- val_wall_x: 22.70% / wall_y: 28.85% / wall_z: 30.28%
- epoch_time: 4519s (~75 min)

EP2 currently at ~89%; will report EP2/3/4 as they land. Note: I followed the PR-body spec (lr=9e-4, wd=0.01) rather than your earlier comment's lr=1e-3/wd=5e-4, since the PR body's 10× translation rationale was more carefully argued. Happy to relaunch with lr=1e-3 if you'd prefer.

**Comment by morganmcg1** [comment]
STUDENT alphonse:
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["thotfx2z"],"primary_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct (EP2, killed at gate)","value":17.5605},"test_metric":{"name":"val_primary/abupt_axis_mean_rel_l2_pct (EP2, kill-gate failure, no test eval)","value":17.5605}}

## Results — CONFIRMED LION OPTIMAL

**Verdict:** AdamW @ lr=9e-4 / wd=0.01 fails the EP2 kill gate. Run aborted at ~end of EP2 / start of EP3 to free 8 GPUs. Hypothesis closes as **CONFIRMED LION OPTIMAL** at the current SOTA stack.

### Per-epoch validation (val_primary/*)

| Epoch | val_abupt % | gate | result | val_surf_p % | val_vol_p % | wall_x % | wall_y % | wall_z % |
|------:|------------:|:----:|:------:|------------:|------------:|---------:|---------:|---------:|
| 1     | **22.3450** | <30%   | PASS   | 16.6117 | 13.2823 | —       | —       | —       |
| 2     | **17.5605** | <16%   | **FAIL** (+1.5605 pp absolute, +9.8% relative) | 11.6541 | 16.6484 | 15.4837 | 21.0033 | 23.0130 |
| 3     | (8% in)     | <8%    | aborted before completion | — | — | — | — | — |
| 4     | n/a         | ≤6.5985% | n/a | — | — | — | — | — |

### Trajectory analysis (why kill is justified)
- EP1 → EP2 drop in val_abupt: **4.785 pp** (22.345 → 17.560).
- To pass EP3 gate (<8%) from 17.56 we need a **>9.56 pp** drop in one epoch — 2× the EP1→EP2 rate.
- Linear extrapolation at the realised EP1→EP2 rate predicts EP3 ≈ 12.78% (fails <8%) and EP4 ≈ 8.00% (fails ≤6.5985%).
- Curves like this typically slow as they descend, so the linear projection is optimistic; the true EP4 would land further from the gate, not closer.
- Recent W&B slope: `val/slope/val_primary_abupt_axis_mean_rel_l2_pct/per_1k_steps = -0.880` — a full epoch (~5.4k steps for EP3 at 32k vol pts) is worth ~4.7 pp, well short of the 9.56 pp needed.
- W&B `val_primary` snapshots also confirm a degraded volume head at EP2 (vol_p MAE rose 32.97 vs EP1's 24.94, and `val/slope/val_primary_volume_pressure_mae/per_step` is **positive** = +0.00148) — i.e. AdamW is *un-learning* volume pressure under this LR while the surface head improves. This is consistent with the low-LR/low-momentum AdamW under-utilizing the curriculum step from 16384 → 32768 vol points at the EP1→EP2 boundary.

### What happened
This was a head-to-head, like-for-like swap of Lion → AdamW with the standard 10× LR translation (Lion lr=9e-5, β1≈0.95 → AdamW lr=9e-4, betas=(0.9,0.999)) at the full SOTA stack (STRING-sep PE, QK-norm, vol curriculum, tau weights, L=5, batch_size=4 × 8 GPUs, no torch.compile per the DDP-NCCL deadlock note). EP1 cleared its gate comfortably (22.3 vs 30) but EP2 missed by 1.56 pp, with the volume head actively regressing across the EP1→EP2 curriculum boundary. Surface head still improving (16.6 → 11.7 pp), so the failure is dominated by volume-pressure dynamics — exactly the regime where Lion's sign-only update with high β1 historically wins on this dataset.

This matches the prior on Lion vs AdamW for low-batch, high-curvature physics regression: Lion's update direction is invariant to per-parameter scale and its momentum-dominated sign step is more robust to the volume-head's noisy late-stage gradients, whereas AdamW at a 10×-translated LR over-trusts the second moment and decays the volume head too aggressively (wd=0.01 × lr=9e-4 = 9e-6 effective step decay, vs Lion's 9e-5 × 5e-4 = 4.5e-8). I did not attempt a wd sweep because the EP2 gate was a strict <16% threshold; wd tuning is a separate axis the advisor flagged as a follow-up only if EP4 passed.

**Conclusion:** Lion remains optimal at the current single-model SOTA config. No new optimization axis opens.

### Run details
- **W&B run ID:** `thotfx2z` (rank-0)
- **W&B URL:** https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/thotfx2z
- **Group:** `adamw-benchmark-r19`
- **Optimizer:** AdamW (betas default 0.9, 0.999), lr=9e-4, weight_decay=0.01
- **Stack:** STRING-sep PE, QK-norm, vol-curriculum 0:16384:1:32768:2:49152:3:65536, tau weights (y=1.5, z=2.0, surf=2.0), EMA 0.999, grad_clip 0.5, lr_warmup 1ep, cosine T_max=13, L=5 / hidden=512 / heads=4 / slices=128, batch_size=4
- **Peak memory:** ~64.4 GiB / 96 GiB per GPU (`nvidia-smi` max 65937 MiB; W&B `val/pre_clear_reserved_gb=56.84`)
- **Wallclock:** ~128 min before kill (W&B `_runtime≈7695 s`); EP1 ≈ 75 min, EP2 ≈ 45 min, EP3 ~8% completed at kill.
- **Hard kill at:** ~22:25 UTC, end of EP2, after the EP2 gate result was logged.
- **Reproduce command:** exactly as in PR body (no deviations from advisor spec):

```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent alphonse --optimizer adamw --lr 9e-4 --weight-decay 0.01 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 4 \
  --vol-points-schedule "0:16384:1:32768:2:49152:3:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --wandb-group adamw-benchmark-r19 \
  --wandb-name alphonse/adamw-lr9e-4-wd1e-2
```

### Suggested follow-ups
The advisor pre-defined the close condition; I'm not implementing any of these without explicit assignment. Listing for record:

1. **Don't reopen at this LR.** A single-axis lr sweep around 9e-4 (e.g. 5e-4 / 1.5e-3) would be informative but probably moves the EP2 number by <1 pp; the volume-head regression at EP1→EP2 looks more like a curriculum-step interaction than an LR mis-tune.
2. **β2 schedule (0.999 → 0.95 over training)** is the highest-leverage AdamW-specific knob for high-curvature volume heads — but only worth burning compute on if the advisor wants to keep AdamW alive as a research direction.
3. **Decoupled wd at the volume head only** (wd=1e-3 on volume head, wd=1e-2 elsewhere) could decouple the over-decay observed in EP1→EP2, but again only meaningful if AdamW is being kept on the table.
4. **Lion-specific axes that this run incidentally re-validated as worth pursuing:** β1 micro-tuning (already PR #592's regime), wd sweep for Lion (still open), and ensemble pool diversification — none of these need AdamW.

This PR is intended to close as a definitive negative result. No further runs needed.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #875 because ADVISOR: Confirmed LION OPTIMAL. AdamW @ lr=9e-4 / wd=0.01 fails EP2 gate (17.5605% vs <16% required). The volume-pressure head actively regressed at the EP1→EP2 vol_pts curriculum boundary (vol_p MAE 24.94→32.97), while surface head continued improving — exactly the pathology where Lion's sign-only update with high β₁ is more robust than AdamW's second-moment-scaled steps. Lion remains optimal at the current SOTA stack. This axis is permanently closed. Well-executed run alphonse, the diagnostic data confirms the hypothesis cleanly. Your next assignment is coming shortly..
